### ISSUE TITLE
typescript >= 2.6 compatibility

### ISSUE THREAD

**Issue by phiresky** [hediet/ts-typed-sql#29]
partial fix:

```patch
commit d10ea00a2e674da8d680559512940b7ecda9a77a
Author: phiresky <phireskyde+git@gmail.com>
Date:   Wed Nov 8 16:38:16 2017 +0100

    fix type error on typescript >= 2.6

diff --git a/src/AST/Expressions.ts b/src/AST/Expressions.ts
index fd4c3c7..6e1030b 100644
--- a/src/AST/Expressions.ts
+++ b/src/AST/Expressions.ts
@@ -291,33 +291,34 @@ export interface ConcreteBinaryExpressionStatic<TLeft extends AnyType, TRight ex
 	new (...args: any[]): ConcreteBinaryExpression<TLeft, TRight, TResult>;
 }
 
-export function ConcreteBinaryExpression<TLeft extends AnyType, TRight extends AnyType, TResult extends AnyType>(
-		symbol: string, resultType: TResult, precedenceLevel: number): ConcreteBinaryExpressionStatic<TLeft, TRight, TResult> {
-	return class extends BinaryOperatorExpression<TLeft, TRight, TResult> {
+export function ConcreteBinaryExpression<TResult extends AnyType>(
+		symbol: string, resultType: TResult, precedenceLevel: number) {
+	return class ConcreteBinaryExpression<TLeft extends AnyType, TRight extends AnyType> extends BinaryOperatorExpression<TLeft, TRight, TResult> {
 		constructor(left: Expression<TLeft>, right: Expression<TRight>) { super(left, right, resultType); }
 		public get operator(): string { return symbol; }
 		public get precedenceLevel(): number { return precedenceLevel; }
 	};
 }
 
-export class AdditionExpression extends ConcreteBinaryExpression<IntegerType, IntegerType, IntegerType>("+", tInteger, 3) {}
-export class SubtractionExpression extends ConcreteBinaryExpression<IntegerType, IntegerType, IntegerType>("-", tInteger, 3) {}
-export class MultiplicationExpression extends ConcreteBinaryExpression<IntegerType, IntegerType, IntegerType>("*", tInteger, 2) {}
-export class DivisionExpression extends ConcreteBinaryExpression<IntegerType, IntegerType, IntegerType>("/", tInteger, 2) {}
-export class ModulusExpression extends ConcreteBinaryExpression<IntegerType, IntegerType, IntegerType>("%", tInteger, 2) {}
+export class AdditionExpression extends ConcreteBinaryExpression<IntegerType>("+", tInteger, 3)<IntegerType, IntegerType> {}
+export class SubtractionExpression extends ConcreteBinaryExpression<IntegerType>("-", tInteger, 3)<IntegerType, IntegerType> {}
+export class MultiplicationExpression extends ConcreteBinaryExpression<IntegerType>("*", tInteger, 2)<IntegerType, IntegerType> {}
+export class DivisionExpression extends ConcreteBinaryExpression<IntegerType>("/", tInteger, 2)<IntegerType, IntegerType> {}
+export class ModulusExpression extends ConcreteBinaryExpression<IntegerType>("%", tInteger, 2)<IntegerType, IntegerType> {}
 
 
-export class EqualsExpression<T extends AnyType> extends ConcreteBinaryExpression<T, T, BooleanType>("=", tBoolean, 4) {}
-export class UnequalsExpression<T extends AnyType> extends ConcreteBinaryExpression<T, T, BooleanType>("!=", tBoolean, 4) {}
-export class GreaterExpression<T extends AnyType> extends ConcreteBinaryExpression<T, T, BooleanType>(">", tBoolean, 4) {}
-export class LessExpression<T extends AnyType> extends ConcreteBinaryExpression<T, T, BooleanType>("<", tBoolean, 4) {}
-export class GreaterOrEqualExpression<T extends AnyType> extends ConcreteBinaryExpression<T, T, BooleanType>(">=", tBoolean, 4) {}
-export class LessOrEqualExpression<T extends AnyType> extends ConcreteBinaryExpression<T, T, BooleanType>("<=", tBoolean, 4) {}
-export class NotLessExpression<T extends AnyType> extends ConcreteBinaryExpression<T, T, BooleanType>("!<", tBoolean, 4) {}
-export class NotGreaterExpression<T extends AnyType> extends ConcreteBinaryExpression<T, T, BooleanType>("!>", tBoolean, 4) {}
+export class EqualsExpression<T extends AnyType> extends ConcreteBinaryExpression<BooleanType>("=", tBoolean, 4)<T, T> {}
 
-export class OrExpression extends ConcreteBinaryExpression<BooleanType, BooleanType, BooleanType>("OR", tBoolean, 7) {}
-export class AndExpression extends ConcreteBinaryExpression<BooleanType, BooleanType, BooleanType>("AND", tBoolean, 6) {}
+export class UnequalsExpression<T extends AnyType> extends ConcreteBinaryExpression<BooleanType>("!=", tBoolean, 4)<T, T> {}
+export class GreaterExpression<T extends AnyType> extends ConcreteBinaryExpression<BooleanType>(">", tBoolean, 4)<T, T> {}
+export class LessExpression<T extends AnyType> extends ConcreteBinaryExpression<BooleanType>("<", tBoolean, 4)<T, T> {}
+export class GreaterOrEqualExpression<T extends AnyType> extends ConcreteBinaryExpression<BooleanType>(">=", tBoolean, 4)<T, T> {}
+export class LessOrEqualExpression<T extends AnyType> extends ConcreteBinaryExpression<BooleanType>("<=", tBoolean, 4)<T, T> {}
+export class NotLessExpression<T extends AnyType> extends ConcreteBinaryExpression<BooleanType>("!<", tBoolean, 4)<T, T> {}
+export class NotGreaterExpression<T extends AnyType> extends ConcreteBinaryExpression<BooleanType>("!>", tBoolean, 4)<T, T> {}
+
+export class OrExpression extends ConcreteBinaryExpression< BooleanType>("OR", tBoolean, 7)<BooleanType, BooleanType> {}
+export class AndExpression extends ConcreteBinaryExpression< BooleanType>("AND", tBoolean, 6)<BooleanType, BooleanType> {}
 
 export class LikeExpression extends Expression<BooleanType> {
 	public get precedenceLevel() { return 7; }

```

**Comment by phiresky** [comment]
relevant: https://github.com/Microsoft/TypeScript/issues/19668

**Comment by hediet** [comment]
Fixed with 35a8dbabd97c78de002a0b9844e0927a577ab639.
